QuantLib/ql/termstructures/yield/fittedbonddiscountcurve.cpp

```diff
@@ -136,36 +144,49 @@ namespace QuantLib {
         Size n = curve_->bondHelpers_.size();
         costFunction_ = shared_ptr<FittingCost>(new FittingCost(this));
         costFunction_->firstCashFlow_.resize(n);
-        weights_ = Array(n);
-        Real squaredSum = 0.0;
-        for (Size i=0; i<curve_->bondHelpers_.size(); ++i) {
-            shared_ptr<Bond> bond = curve_->bondHelpers_[i]->bond();
-
-            Leg leg = bond->cashflows();
-            Real cleanPrice = curve_->bondHelpers_[i]->quote()->value();
-            
-            Date bondSettlement = bond->settlementDate();
-            Rate ytm = BondFunctions::yield(*bond, cleanPrice,
+        if(weights_.size() != n)
+        {
+            weights_ = Array(n);
+            Real squaredSum = 0.0;
+            for (Size i=0; i<curve_->bondHelpers_.size(); ++i) {
+                shared_ptr<Bond> bond = curve_->bondHelpers_[i]->bond();
+
+                Leg leg = bond->cashflows();
+                Real cleanPrice = curve_->bondHelpers_[i]->quote()->value();
+
+                Date bondSettlement = bond->settlementDate();
+                Rate ytm = BondFunctions::yield(*bond, cleanPrice,
                                             yieldDC, yieldComp, yieldFreq,
                                             bondSettlement);
 
-            Time dur = BondFunctions::duration(*bond, ytm,
+                Time dur = BondFunctions::duration(*bond, ytm,
                                                yieldDC, yieldComp, yieldFreq,
                                                Duration::Modified,
                                                bondSettlement);
-            weights_[i] = 1.0/dur;
-            squaredSum += weights_[i]*weights_[i];
-
-            const Leg& cf = bond->cashflows();
-            for (Size k=0; k<cf.size(); ++k) {
-                if (!cf[k]->hasOccurred(bondSettlement, false)) {
-                    costFunction_->firstCashFlow_[i] = k;
-                    break;
+                weights_[i] = 1.0/dur;
+                squaredSum += weights_[i]*weights_[i];
+
+                const Leg& cf = bond->cashflows();
+                for (Size k=0; k<cf.size(); ++k) {
+                    if (!cf[k]->hasOccurred(bondSettlement, false)) {
+                        costFunction_->firstCashFlow_[i] = k;
+                        break;
+                    }
                 }
             }
+            weights_ /= std::sqrt(squaredSum);
         }
-        weights_ /= std::sqrt(squaredSum);
+    }
+
 
+    DiscountFactor FittedBondDiscountCurve::FittingMethod::discountFunction(const Array& x,
```

Comment:
This is back to being pure virtual; the implementation should be deleted (it no longer compiles anyway).